### ISSUE TITLE
bug: text generation result is different for different operating systems (macOS, windows, and Linux)

### ISSUE THREAD

**Issue by l1xiangyi** [pytorch/pytorch#111315]
### 🐛 Describe the bug

# Repo of the project where the bug is spotted:
https://github.com/l1xiangyi/GPT2/tree/main

# What is the problem
The problem occurs in [main.py line 89](https://github.com/l1xiangyi/GPT2/blob/b0a759b6a6fed6dc87d2f450f0cb9088e81937d1/main.py#L89) where I try to generate texts based on a NeruIPS dataset. This line calls the sample function that generates subsequent sentences based on the `start_text`. However the texts generated on macOS, Windows, and Linux are very different even though the seed is fixed in the [`config.yml`](https://github.com/l1xiangyi/GPT2/blob/main/config.yml). 

- macOS result: [`sumbit/samples.txt`](https://github.com/l1xiangyi/GPT2/blob/main/submit/samples.txt)
- Windows result: [`sumbit/sample_windows.txt`](https://github.com/l1xiangyi/GPT2/blob/main/submit/sample_windows.txt)
- Windows result: [`sumbit/sample_linux.txt`](https://github.com/l1xiangyi/GPT2/blob/main/submit/sample_linux.txt) 

Strange words like "Zombie" pop up a lot in the Windows and Linux generated texts. 

# How to reproduce this problem

## Linux
The simplest way is go create a new Colab notebook and run the following commands:
```
!git clone https://github.com/l1xiangyi/gpt2/
%cd gpt2
!python -m pip install -r requirements.txt
!python main.py
```
Then check the `samples.txt` in the `submit` folder. 

## Windows
Similar steps to the Linux version. Just drop the special sign before each commands:
```
git clone https://github.com/l1xiangyi/gpt2/
cd gpt2
python -m pip install -r requirements.txt
python main.py
```

### Versions

# Versions:
## macOS:
Ran
```
python3 collect_env.py
```
Results: 
```
Collecting environment information...
PyTorch version: 2.1.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.0 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.0.40.1)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.6 (default, Aug 11 2023, 19:44:49)  [Clang 15.0.0 (clang-1500.0.40.1)] (64-bit runtime)
Python platform: macOS-14.0-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Max

Versions of relevant libraries:
[pip3] numpy==1.25.2
[pip3] torch==2.1.0
[conda] Could not collect
```

## Windows
Ran
```
!python collect_env.py
```
Result
```
Collecting environment information...
PyTorch version: 2.1.0+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Home China
GCC version: (x86_64-win32-seh-rev0, Built by MinGW-W64 project) 8.1.0
Clang version: Could not collect
CMake version: version 3.25.1
Libc version: N/A

Python version: 3.8.18 (default, Sep 11 2023, 13:39:12) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19045-SP0
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1660 Ti
Nvidia driver version: 497.09
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:

Revision=

Versions of relevant libraries:
[pip3] numpy==1.24.4
[pip3] torch==2.1.0
[conda] No relevant packages

Process finished with exit code 0
```

## Linux:
Ran
```
!python collect_env.py
```
Result:
```
2023-10-14 21:56:43 (12.8 MB/s) - ‘collect_env.py’ saved [21737/21737]

Collecting environment information...
PyTorch version: 2.0.1+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.27.6
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.120+-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          2
On-line CPU(s) list:             0,1
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) CPU @ 2.20GHz
CPU family:                      6
Model:                           79
Thread(s) per core:              2
Core(s) per socket:              1
Socket(s):                       1
Stepping:                        0
BogoMIPS:                        4399.99
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       32 KiB (1 instance)
L1i cache:                       32 KiB (1 instance)
L2 cache:                        256 KiB (1 instance)
L3 cache:                        55 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0,1
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Vulnerable; SMT Host state unknown
Vulnerability Meltdown:          Vulnerable
Vulnerability Mmio stale data:   Vulnerable
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:        Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Vulnerable

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==2.0.1+cu118
[pip3] torchaudio==2.0.2+cu118
[pip3] torchdata==0.6.1
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.15.2
[pip3] torchvision==0.15.2+cu118
[pip3] triton==2.0.0
[conda] Could not collect
```

**Comment by janeyx99** [comment]
I am not sure pytorch/pytorch is the best place to discuss this bug as it is not clear the bug is pytorch related--you may get more useful insight from asking our forum or by isolating a smaller repro. I do not know much about seed generation but I could conceive that the logic may be different on different platforms. That said, I would not expect the model to be do dependent on the original seed, so there may be bigger problems elsewhere.

**Comment by l1xiangyi** [comment]
Thank you Jane for explanation. We tried debugging the seed generation at first though but it does not seem to be problem with seed mechanisms. We also tried eliminating difference between line break symbols between mac and windows but also to no avail. Thank you for replying, and if it's not problem with the PyTorch framework then it possible is related to the pre-trained GPT model itself. Thank you for pointing out the this bug is not clearly PyTorch related. It's just we tried many possibilities outside of PyTorch so I just posted it here. You can close my issue until I could repro it in a smaller scope.

**Comment by janeyx99** [comment]
Okay! Closing but ofc feel free to reopen when you can identify that it's a torch issue~
